server Stat call returns incorrect error for some non-existent Windows file paths

(Apologies if I get some of the details wrong here, as I'm just digging into filesystem syscalls as a result of this bug.  This is all foreign territory for me.  But I do feel pretty confident about the fix I have)

---

`os.Stat` on Windows returns `ENOENT` if the requested file or directory does not exist, but `ENOTDIR` if the requested file is a child of a non-existent directory.  For example, given the code:

``` go
_, err1 := os.Stat("/doesnotexist")
fmt.Printf("%#v\n", err1)

_, err2 := os.Stat("/doesnotexist/a/b")
fmt.Printf("%#v\n", err2)
```

On windows, this prints:

``` go
&os.PathError{Op:"CreateFile", Path:"/doesnotexist", Err:0x2}      // ENOENT
&os.PathError{Op:"CreateFile", Path:"/doesnotexist/a/b", Err:0x3}  // ENOTDIR
```

Compared to linux, which treats these both the same:

``` go
&os.PathError{Op:"stat", Path:"/doesnotexist", Err:0x2}      // ENOENT
&os.PathError{Op:"stat", Path:"/doesnotexist/a/b", Err:0x2}  // ENOENT
```

sftp's server implementation maps these errors to ssh status codes in [statusFromError](https://github.com/pkg/sftp/blob/ec5b80d98efe8089fc64b9acafabc0af1d1d70b2/server.go#L568).  In the case of `os.PathError`, this calls [translateErrno](https://github.com/pkg/sftp/blob/ec5b80d98efe8089fc64b9acafabc0af1d1d70b2/server.go#L555) which properly translates the `ENOENT` to `sshFxNoSuchFile`, but the `ENOTDIR` gets left as a generic `sshFxFailure`.

This means that when the response gets [converted back into a standard error](https://github.com/pkg/sftp/blob/ec5b80d98efe8089fc64b9acafabc0af1d1d70b2/client.go#L1313) in an sftp client, it is not seen as a not exists error.  Here is a simple reproduction case:

``` go
func TestStatNonExistent(t *testing.T) {
	cPipe, sPipe := net.Pipe()
	srv, _ := sftp.NewServer(sPipe)
	go srv.Serve()
	client, _ := sftp.NewClientPipe(cPipe, cPipe)

	for _, file := range []string{"/doesnotexist", "/doesnotexist/a/b"} {
		_, err := client.Stat(file)
		if err == nil || !os.IsNotExist(err) {
			t.Errorf("expected 'does not exist' err for file %q.  got: %v", file, err)
		}
	}
}
```

This works fine on Linux, but on Windows returns:

``` sh
--- FAIL: TestStatNonExistent (0.00s)
    c:\Users\willnorris\src\sftptest\sftp_test.go:30: expected 'does not exist' err for file "/doesnotexist/a/b".  
got: sftp: "CreateFile /doesnotexist/a/b: The system cannot find the path specified." (SSH_FX_FAILURE)
```

It's tempting to just have `translateErrno` handle `ENOTDIR` as well, but golang/go#18974 suggests that is not always a good idea.  I'm not 100% sure if the use cases Ian talks about in that bug apply here, but it's not necessary anyway.  As of [a919b76](https://github.com/golang/go/commit/a919b760378343a211719991a9e0b1b819ac9f3d), `os.IsNotExist` properly identifies `ENOTDIR` on windows, so it's as simple as checking `os.IsNotExist` inside of `statusFromError`, and mapping it to `sshFxNoSuchFile`.  I've verified this resolves the reproduction case above, and will followup with that in a PR shortly.

(For what it's worth, I found this while trying to debug [this Windows test failure](https://github.com/willnorris/perkeep/runs/1104803999?check_suite_focus=true#step:6:800) in Perkeep)

--- a/server.go
+++ b/server.go
@@ -589,24 +589,24 @@ func statusFromError(p ider, err error) sshFxpStatusPacket {
 	ret.StatusError.Code = sshFxFailure
 	ret.StatusError.msg = err.Error()
 
+	if os.IsNotExist(err) {
+		ret.StatusError.Code = sshFxNoSuchFile
+		return ret
+	}
+
 	switch e := err.(type) {
 	case syscall.Errno:
 		ret.StatusError.Code = translateErrno(e)
 	case *os.PathError:
 		debug("statusFromError,pathError: error is %T %#v", e.Err, e.Err)
-		if os.IsNotExist(err) {
-			ret.StatusError.Code = sshFxNoSuchFile
-		} else if errno, ok := e.Err.(syscall.Errno); ok {
+		if errno, ok := e.Err.(syscall.Errno); ok {
 			ret.StatusError.Code = translateErrno(errno)
 		}
 	case fxerr:
 		ret.StatusError.Code = uint32(e)
 	default:
-		switch e {
-		case io.EOF:
+		if e == io.EOF {
 			ret.StatusError.Code = sshFxEOF
-		case os.ErrNotExist:
-			ret.StatusError.Code = sshFxNoSuchFile
 		}
 	}
 

--- a/server.go
+++ b/server.go
@@ -594,7 +594,9 @@ func statusFromError(p ider, err error) sshFxpStatusPacket {
 		ret.StatusError.Code = translateErrno(e)
 	case *os.PathError:
 		debug("statusFromError,pathError: error is %T %#v", e.Err, e.Err)
-		if errno, ok := e.Err.(syscall.Errno); ok {
+		if os.IsNotExist(err) {
+			ret.StatusError.Code = sshFxNoSuchFile
+		} else if errno, ok := e.Err.(syscall.Errno); ok {
 			ret.StatusError.Code = translateErrno(errno)
 		}
 	case fxerr:

--- a/server_test.go
+++ b/server_test.go
@@ -354,7 +354,7 @@ func TestStatNonExistent(t *testing.T) {
 
 	for _, file := range []string{"/doesnotexist", "/doesnotexist/a/b"} {
 		_, err := client.Stat(file)
-		if err == nil || !os.IsNotExist(err) {
+		if !os.IsNotExist(err) {
 			t.Errorf("expected 'does not exist' err for file %q.  got: %v", file, err)
 		}
 	}

--- a/server_test.go
+++ b/server_test.go
@@ -344,3 +344,18 @@ func TestOpenStatRace(t *testing.T) {
 	os.Remove(tmppath)
 	checkServerAllocator(t, server)
 }
+
+// Ensure that proper error codes are returned for non existent files, such
+// that they are mapped back to a 'not exists' error on the client side.
+func TestStatNonExistent(t *testing.T) {
+	client, server := clientServerPair(t)
+	defer client.Close()
+	defer server.Close()
+
+	for _, file := range []string{"/doesnotexist", "/doesnotexist/a/b"} {
+		_, err := client.Stat(file)
+		if err == nil || !os.IsNotExist(err) {
+			t.Errorf("expected 'does not exist' err for file %q.  got: %v", file, err)
+		}
+	}
+}

